### PR TITLE
Add a device type filter to the Transactions list page

### LINKED ISSUE #5756: Add Tap to Pay device type filter on transactions list view
### Description
In the Transactions list view (Payments > Transactions) add a new advanced filter `Device type` to filter the Tap to Pay transactions. The new filter should accept two values: `iPhone` and `Android`.

### PR DESCRIPTION
**Pull request by mordeth**
Fixes #5756

#### Changes proposed in this Pull Request
Adding a Tap to Pay device source filter to the Transactions list page.

Screenshot 2023-05-15 at 14 23 55

#### Testing instructions
- One the server use 3371-gh-Automattic/woocommerce-payments-server PR
- Create orders paid with Tap to Pay.
- Navigate to Payment -> Transactions.
- Add a Device Type advanced filter.
- Verify the transactions in the list are only the Tap to Pay.
- Verify the transactions summary and pagination are working correctly.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6295` or branch name `add/ttp-transactions-list-filter` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/ttp-transactions-list-filter) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 4311000cd319905cc9b8d365870385e642d82bff
- Build time: 2023-06-15 05:47:24 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/transactions/strings.ts [hunk 1]
```diff
@@ -22,3 +22,9 @@ export const displayType = {
 	financing_payout: __( 'Loan dispersement', 'woocommerce-payments' ),
 	financing_paydown: __( 'Loan repayment', 'woocommerce-payments' ),
 };
+
+// Mapping of transaction device type string.
+export const sourceDevice = {
+	android: __( 'Android', 'woocommerce-payments' ),
```
Comment (anu-rock) [inline client/transactions/strings.ts]: As we discussed during our demo, we should remove this option (and update all tests) as we don't yet support Android. Feel free to check with Design too.

File: includes/core/server/request/class-list-transactions.md [hunk 2]
```diff
@@ -24,6 +24,8 @@ The `WCPay\Core\Server\Request\List_Transactions` class is used to construct the
 | `store_currency_is`        | `set_store_currency_is( string $currency )`                |     -     |    -     |       -       |
 | `type_is`                  | `set_type_is( string $type_is )`                           |     -     |    -     |       -       |
 | `type_is_not`              | `set_type_is_not( string $type_is_not )`                   |     -     |    -     |       -       |
+| `source_device_is`         | `set_source_device_is( string $source_device_is )`         |     -     |    -     |       -       |
```
Comment (anu-rock) [inline includes/core/server/request/class-list-transactions.md:27]: TIL that we maintain easy-to-read in-code docs for server requests 🙌

**Review by anu-rock — COMMENTED** [review]
The changes are streamlined and lovely. Here's what I checked in my first round of review:

| Parameter  | Result |
| ------------- | ------------- |
| Approach  | 👍  Almost good (left a couple of comments) |
| Implementation  | ✅ All good |
| Tests  | ✅ All good |
| Naming  | ✅ All good |
| Typing  | ✅ All good |
| Changelog  | ✅ All good |
| Testing instructions  | 👍  Almost good (how to create TTP orders could be expanded) |
| Manual testing  | ❌ Failed. See my comment on 3371-gh-Automattic/woocommerce-payments-server. |
| Edge cases  | ❌ Could not test because of the above error |

**Comment by anu-rock** [comment]
Second round:

| Parameter  | Result |
| ------------- | ------------- |
| Manual testing: "is iphone" filter  | ✅ All good |
| Manual testing: "is not iphone" filter  | ❌ No transactions displayed. Should it not show all transactions except TTP with iPhone? |
| Manual testing: reload URL with filter [query string](http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Ftransactions&filter=advanced&source_device_is=ios)  | ✅ All good |
| Edge case: "is iphone" filter when there is no TTP transaction  | ✅ All good |
| Edge case: "is not iphone" filter when there is no TTP transaction  | ❌ No transactions displayed. Should it not show all transactions except TTP with iPhone? |

File: client/transactions/strings.ts [hunk 3]
```diff
@@ -22,3 +22,9 @@ export const displayType = {
 	financing_payout: __( 'Loan dispersement', 'woocommerce-payments' ),
 	financing_paydown: __( 'Loan repayment', 'woocommerce-payments' ),
 };
+
+// Mapping of transaction device type string.
+export const sourceDevice = {
+	// android: __( 'Android', 'woocommerce-payments' ),
```
Comment (anu-rock) [inline client/transactions/strings.ts]: IMO, we should remove the commented-out code and reintroduce it later when appropriate.

Comment (anu-rock) [reply client/transactions/strings.ts]: We should also update the tests and snapshots to remove Android.

**Review by anu-rock — APPROVED** [review]
As discussed in our team call, the behavior I observed for the "is not" filter in my edge case comments is, though valid, how the component works for all other filters too. @mordeth to start a separate discussion about it and create a new issue if required.

Pre-approving with a couple of minor comments.

**Review by dpaun1985 — APPROVED** [review]
✅  LGTM

**Pull request merged**